src/Mvc/Mvc.Testing/src/WebApplicationFactory.cs

```diff
@@ -164,13 +166,38 @@ private void EnsureServer()
             _server = CreateServer(builder);
         }
 
-        private void SetContentRoot(IWebHostBuilder builder)
-        {
+        private void SetContentRoot(IWebHostBuilder builder) {
             if (SetContentRootFromSetting(builder))
             {
                 return;
             }
 
+            // Where do we store the file that we read the JSON config from
+            var fromFile = File.Exists("AppManifest.json");
+            var contentRoot = fromFile ? GetContentRootFromFile("AppManifest.json") : GetContentRootFromAssembly();
+
+            if (contentRoot != null)
+            {
+                builder.UseContentRoot(contentRoot);
+            }
+            else
+            {
+                builder.UseSolutionRelativeContentRoot(typeof(TEntryPoint).Assembly.GetName().Name);
+            }
+        }
+
+        private string GetContentRootFromFile(string file) {
+            var serializer = new DataContractJsonSerializer(typeof(Dictionary<string, string>), new DataContractJsonSerializerSettings
```

Comment:
Use S.T.Json. The only reason you wouldn't want to use this the DataContract serializer is because you're forced to.

Reply:
Also formatting

Reply:
We ran into a weird issue while we were pairing on this where `S.T.Json` could not parse the JSON that was generated from the task. We decided to switch to using the `DataContractJsonSerializer` since it seemed like `S.T.Json` wasn't compatible with whatever `DataContract` serializer was writing.

We had mentioned changing the task to use `S.T.Json` but I believe we opted not to do that for a reason I can't recall now but @javiercn or @HaoK might remeber.

Reply:
@javiercn said to trust him that we need to do it :)

Reply:
Can you include one the files that DataContract produces?

Reply:
Here it is: https://gist.github.com/captainsafia/9c61cdd4e410b43fcda6d122f5b48715

We tinkered around with changing the generic on the Deserialize call when we were using the S.T.Json deserializer. We tried `Dictionary<string, string>`, `JsonElement`, and `IDictionary<string, string>` and all of them resulted in parsing failures.

Reply:
Followed up with @pranavkm to understand the trade-offs here. We can use `S.T.Json` in the build tasks but we would have to do some extra work to ensure that the `S.T.Json` dependency is bundled with the tasks. Given that this is a small infrastructure change, we probably want to avoid any work that isn't necessary here.

Then there's the question of why `S.T.Json` has trouble reading a file generated by `DataContractJsonSerializer`. I'll open an issue on the runtime repo for this.

Reply:
I mean at the end of the day we really don't care about the format, we could just write it out using strings on a line too...

Reply:
Json is nice if we end up needing more complicated things in the manifest, I'm a bit surprised that the only thing we seem to need for this is content root mappings...

Reply:
Btw, I tried using STJ to read the contents of the gist and it looks fine. 

Unrelated, one of the things we need to be cognizant of is that having an MSBuild target that we build as part of AspNetCore and also consume as part of the same build runs the risk of being locked by Visual Studio.

Reply:
> Btw, I tried using STJ to read the contents of the gist and it looks fine.

Figured out what it was. It seems the behavior is distinct when you pass a file path vs a `FileStream` to the deserialize API.

Reply:
Perhaps its some kind of first byte marker that isn't dealt with automatically by file system vs stream?  I vaguely recall running into BOM issues like this in the past... especially since you are non windows maybe that is a possibility?

Reply:
lol, turns out it is actually because `S.T.Json` doesn't have a deserialize API that takes a filepath as an argument.

Reply:
So it was trying to deserialize the file path as json?  whoops haha